common/src/java/org/apache/hadoop/hive/ql/log/PerfLogger.java

```diff
@@ -129,9 +133,31 @@ public static void setPerfLogger(PerfLogger resetPerfLogger) {
    * @param method method or ID that identifies this perf log element.
    */
   public void perfLogBegin(String callerName, String method) {
+    perfLogBegin(callerName, method, null);
+  }
+
+  /**
+   * Call this function when you start to measure time spent by a piece of code.
+   * @param callerName the logging object to be used.
+   * @param method method or ID that identifies this perf log element.
+   * @param additionalInfo particular step within the method that is being measured
+   */
+  public void perfLogBegin(String callerName, String method, String additionalInfo) {
     long startTime = System.currentTimeMillis();
-    startTimes.put(method, Long.valueOf(startTime));
-    LOG.debug("<PERFLOG method={} from={}>", method, callerName);
+    String key = method;
+    if (additionalInfo != null) {
+      key = key + " - " + additionalInfo;
+    }
+    startTimes.put(key, Long.valueOf(startTime));
+    if (LOG.isDebugEnabled()) {
+      StringBuilder sb = new StringBuilder("<PERFLOG method=").append(method);
+      sb.append(" from=").append(callerName);
+      if (additionalInfo != null) {
+        sb.append(" ").append(additionalInfo);
+      }
+      sb.append(">");
+      LOG.debug(sb.toString());
+    }
     beginMetrics(method);
```

Comment:
It seems that the `PerfLogger` already tracks some metrics for each call on begin/end. Did you check if its possible to exploit the metrics to provide these kind of cumulative optimization times? Why do we keep these metrics are they useful in any way?

`PerfLogger` is called in many places and also gets heavier and more involved every time. I want to ensure that we are not adding new information and logic that we don't already have somewhere.